### DISCUSSION TITLE
Customizable Outliner

### DISCUSSION THREAD

**Discussion by jossh-h** [Autodesk/maya-usd#2471] [Ideas] [answered]
Hi there,

We are exploring if and how we can switch over from our in house USD outliner to using the Maya native outliner.
Some of the features we would like are:
 1) Filter the outliner based on our own criteria, for example prim type, metadata
 2) Ability to control how a prim is displayed, grey-out/strikethrough/icon/text color
 3) Additional columns  with additional row information.

**1) Filter the outliner based on our own criteria, for example prim type and/or metadata**
We have cases where we want to hide prims from the outliner. If we were dealing with Maya data we could have used the [outlinerEditor](https://help.autodesk.com/cloudhelp/2023/ENU/Maya-Tech-Docs/CommandsPython/outlinerEditor.html) and [itemFilter](https://help.autodesk.com/cloudhelp/2023/ENU/Maya-Tech-Docs/CommandsPython/itemFilter.html) commands, but these commands don't seem to be fully UFE aware. I did see the work done in this [PR](https://github.com/Autodesk/maya-usd/pull/773) but that seems to be hard coded to toggle the display or active/inactive prims. 
Is there any plan to allow this to be easily extensible possibly through a plugin or callback?
 
**2) Control how prims are displayed,  grey-out/strike-through/icon/text color**
In some cases we want to give our artists some visual feeback about a particular prim, this may be a change of icon, grey-out or maybe even text color. 

Has there been any talks about customizing how a prim is shown in the outliner?

**3) Ability to extend the outliner with extra columns with additional row information.**
Not sure if it is feasible to allow this in the Maya outliner but curious how others have tackled this, we will likely create a qt gui with additional information.

Ideally we would have control over what (filtering) and how (grey-out/strikethrough/icon) prims are displayed in the outliner

Thanks,
Josh

**Comment by ppt-adsk** [comment, marked as answer]
Hi @jossh-h, let me try to answer your questions:

1. Outliner hierarchy filtering: the pull request you were looking at is the right one, and is not hard-coded: I'm hopeful we have that one covered to some extent.  Have a look at the UFE Hierarchy interface child filtering capability:
https://help.autodesk.com/view/MAYAUL/2023/ENU/?guid=Maya_SDK_ufe_ref_class_ufe_1_1_hierarchy_html
This allows the concrete Hierarchy interface to implement filtering on the children it reports.  We use that here in the USD implementation of the Hierarchy interface to control whether inactive prims are shown or not:
https://github.com/Autodesk/maya-usd/blob/c8291ec838debacc71de5994ec3dcb845a6a52ea/lib/mayaUsd/ufe/UsdHierarchyHandler.cpp#L62
https://github.com/Autodesk/maya-usd/blob/c8291ec838debacc71de5994ec3dcb845a6a52ea/lib/mayaUsd/ufe/UsdHierarchy.cpp#L198
If this is something that can be made to work in a proof of concept, then we can figure out how to make it happen without invasively changing the core maya-usd code, possibly with a Decorator pattern implementation:
https://en.wikipedia.org/wiki/Decorator_pattern

2. Prim display control: here's another one I hope we can help with.  Look at the UIInfoHandler::treeViewCellInfo() and treeViewIcon() methods.  Dynamic icon customization is supported; have a look at my final answer [here](https://github.com/Autodesk/maya-usd/discussions/2499).  The UFE CellInfo struct
https://help.autodesk.com/view/MAYAUL/2023/ENU/?guid=Maya_SDK_ufe_ref_struct_ufe_1_1_cell_info_html
allows for customization of strike-out, text foreground color, and text background color.  Again, once a proof of concept is working, we can figure out how to deploy it.

3. Customizable extra Outliner columns at each row: currently not possible.  This would be a more ambitious project for us.

Hope this helps.  Many thanks to @seando-adsk for considerable help.

**Reply by jossh-h** [reply]
Thanks for looking at this @ppt-adsk @seando-adsk .
We can look at creating a proof of concept for the outliner filtering and will touch base with @dj-mcg on https://github.com/Autodesk/maya-usd/discussions/2499 incase there is a double up for the work on custom icons :)
